program/src/processor/harvest_sync_rewards.rs

```diff
@@ -0,0 +1,232 @@
+        ctx.accounts.sol_stake.key == &sol_staker_stake.sol_stake,
+        StakeError::IncorrectSolStakeAccount,
+        "sol stake"
+    );
+
+    require!(
+        ctx.accounts.sol_stake_view_program.key == &paladin_sol_stake_view_program_client::ID,
+        ProgramError::IncorrectProgramId,
+        "invalid sol stake view program"
+    );
+
+    GetStakeActivatingAndDeactivatingCpiBuilder::new(ctx.accounts.sol_stake_view_program)
+        .stake(ctx.accounts.sol_stake)
+        .stake_history(ctx.accounts.sysvar_stake_history)
+        .invoke()?;
+
+    let (_, return_data) = get_return_data().ok_or(ProgramError::InvalidAccountData)?;
+    let mut stake_state_data =
+        bytemuck::try_from_bytes::<GetStakeActivatingAndDeactivatingReturnData>(&return_data)
+            .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    let delegated_vote = stake_state_data.delegated_vote.get();
+    // If we have the correct sol stake account, but the delegated vote is not the same
+    // as the validator vote, then the stake is not delegated to the SOL staker stake
+    // account's validator vote anymore; we will clear the lamports amount in this case.
+    if delegated_vote.is_none()
+        || delegated_vote != Some(sol_staker_stake.delegation.validator_vote)
+    {
+        stake_state_data = bytemuck::from_bytes(&EMPTY_RETURN_DATA);
+    }
+
+    // Check whether the SOL stake is ouf-of-sync.
+
+    let stake_amount = u64::from(stake_state_data.effective)
+        .checked_add(u64::from(stake_state_data.activating))
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    if stake_amount != sol_staker_stake.lamports_amount {
+        msg!("SOL stake is out-of-sync sync");
+
+        // When the SOL stake is out-of-sync, we need to:
+        //
+        //   1. Calculate the rewards earned by the SOL staker stake account, so we
+        //      can determine how much fees the searcher should receive. Searcher fee
+        //      is the min of (rewards available, config search fee)
+        //
+        //   2. Update the `last_seen_stake_rewards_per_token` on the SOL staker stake
+        //      account to deduct the searcher fee. This will decrease the rewards per
+        //      token for the current SOL staker stake account to take into consideration
+        //      the searcher reward.
+        //
+        //   3. Transfer the searcher fee to the destination account
+        //
+        //   4. Update the SOL staker and validator stake accounts' lamports amount
+        //
+        // NOTE: steps 2 and 3 only happen when the searcher fee is greater than 0.
+
+        // searcher rewards
+        let rewards = calculate_eligible_rewards(
+            u128::from(config.accumulated_stake_rewards_per_token),
+            u128::from(
+                sol_staker_stake
+                    .delegation
+                    .last_seen_stake_rewards_per_token,
+            ),
+            sol_staker_stake.delegation.amount,
```

Comment:
We are not capping the rewards here, so we have funds to pay the sync rewards. In a scenario where `stake_amount` is `0` but there are rewards available, the sync rewards are paid from the excess.

Reply:
There's some weird game theory here then. You can keep your PAL account over-staked (PAL amount much higher than SOL) and use this to get extra rewards that you otherwise wouldn't get if you did normal syncing.

Reply:
But I think you're right, this probably makes the most sense -- oh well!